include/jemalloc/internal/rtree_inlines.h

```diff
@@ -170,103 +162,86 @@ rtree_subtree_read(tsdn_t *tsdn, rtree_t *rtree, unsigned level,
 JEMALLOC_ALWAYS_INLINE rtree_elm_t *
 rtree_elm_lookup(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx,
     uintptr_t key, bool dependent, bool init_missing) {
-	uintptr_t subkey;
-	unsigned start_level;
-	rtree_elm_t *node;
-
 	assert(!dependent || !init_missing);
 
-	if (dependent || init_missing) {
-		if (likely(rtree_ctx->valid)) {
-			if (key == rtree_ctx->key) {
-				return rtree_ctx->elms[rtree->height];
-			} else {
-				unsigned no_ctx_start_level =
-				    rtree_start_level(rtree, key);
-				unsigned ctx_start_level;
-
-				if (likely(no_ctx_start_level <=
-				    rtree_ctx->start_level && (ctx_start_level =
-				    rtree_ctx_start_level(rtree, rtree_ctx,
-				    key)) >= rtree_ctx->start_level)) {
-					start_level = ctx_start_level;
-					node = rtree_ctx->elms[ctx_start_level];
-				} else {
-					start_level = no_ctx_start_level;
-					node = init_missing ?
-					    rtree_subtree_read(tsdn, rtree,
-					    no_ctx_start_level, dependent) :
-					    rtree_subtree_tryread(rtree,
-					    no_ctx_start_level, dependent);
-					rtree_ctx->start_level =
-					    no_ctx_start_level;
-					rtree_ctx->elms[no_ctx_start_level] =
-					    node;
+	/* Search the cache. */
+	uintptr_t leafkey = rtree_leafkey(rtree, key);
+	if (likely(key != 0)) {
+		if (likely(rtree_ctx->cache[0].leafkey == leafkey)) {
+			rtree_elm_t *leaf = rtree_ctx->cache[0].leaf;
+			if (likely(leaf != NULL)) {
+				uintptr_t subkey = rtree_subkey(rtree, key,
+				    rtree->height-1);
+				return &leaf[subkey];
+			}
+		}
+		/*
+		 * Search the remaining cache elements, and on success move the
+		 * matching element to the front.
+		 */
+		for (unsigned i = 1; i < RTREE_CTX_NCACHE; i++) {
+			if (likely(rtree_ctx->cache[i].leafkey == leafkey)) {
+				rtree_elm_t *leaf = rtree_ctx->cache[i].leaf;
+				if (likely(leaf != NULL)) {
+					/* Reorder. */
+					memmove(&rtree_ctx->cache[1],
```

Comment:
Probably worth having a #ifdef RTREE_CTX_NCACHE == 2, do an assignment directly instead of memmove? I'm not quite optimistic with all cases / compilers there.